edgexfoundry/device-gpio
# refactor(snap): edgex-snap-hooks related upgrade

This PR upgrades snap hooks to [2.4.1 ](https://github.com/canonical/edgex-snap-hooks/releases/tag/v2.4.1)with new function and refactor:
- Combine install hook and configure hook
- Use library function for copying directory
- Remove export bash command to process valid .env files
- Use log package for logging
- Use library function for autostart processing

Signed-off-by: Mengyi Wang <mengyi.wang@canonical.com>







**If your build fails** due to your commit message not passing the build checks, please review the guidelines here: https://github.com/edgexfoundry/device-gpio/blob/main/.github/Contributing.md

## PR Checklist
Please check if your PR fulfills the following requirements:

- [x] I have fully tested (add details below) this the new feature or bug fix (if not, why?)  [gpio test suite](https://github.com/canonical/edgex-snap-testing/tree/main/test/suites/device-gpio) passed with `FULL_CONFIG_TEST=true`
  <link to docs PR>

## Testing Instructions

1. Install and validate all config files have been copied in right place:
```
tree /var/snap/edgex-device-gpio/current/config
.
└── device-gpio
    └── res
        ├── configuration.toml
        ├── devices
        │   ├── device.custom.gpio.toml
        │   └── device.led.gpio.toml
        └── profiles
            ├── device.custom.gpio.yaml
            └── device.led.gpio.yaml

4 directories, 6 files
```
2. Setup and start:
```
sudo snap set edgex-device-gpio app-options=true
sudo snap set edgex-device-gpio config.writable-loglevel=DEBUG
sudo snap set edgex-device-gpio config.service-startupmsg="#test start up message#"
sudo snap start edgex-device-gpio
```
3. Check the .env file is valid now without `export` command:
```
cat /var/snap/edgex-device-gpio/current/config/device-gpio/res/device-gpio.env 

SERVICE_STARTUPMSG="#test start up message#"
WRITABLE_LOGLEVEL="DEBUG"
```
4. Validate env vars in `.env` have been exported successfully:
```
sudo snap logs -f edgex-device-gpio

msg="#test start up message#"
level=DEBUG msg="xxx"
```
5. Validate autostart via snap option works:
```
$ sudo snap set edgex-device-gpio autostart=true
$ snap services edgex-device-gpio
Service                        Startup  Current  Notes
edgex-device-gpio.device-gpio  enabled  active   -

$ sudo snap set edgex-device-gpio autostart=false
$ snap services edgex-device-gpio
Service                        Startup   Current   Notes
edgex-device-gpio.device-gpio  disabled  inactive  -
```

## New Dependency Instructions (If applicable)
